src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3WebtransportSession.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Quic;
+using System.Runtime.Versioning;
+using System.Text;
+using System.Net.Http.Headers;
+using System.Net.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System.IO;
+using System.Threading.Channels;
+
+namespace System.Net.Http
+{
+    [SupportedOSPlatform("windows")]
+    [SupportedOSPlatform("linux")]
+    [SupportedOSPlatform("macos")]
+    [UnsupportedOSPlatform("browser")]
+
+    public class Http3WebtransportSession : IAsyncDisposable, IDisposable
+    {
+        private readonly QuicStream _connectStream;
+        private static HttpMessageInvoker? s_defaultInvoker;
+
+        public long Id
+        {
+            get { return _connectStream.Id; }
+        }
+
+        private readonly Channel<QuicStream> _incomingStreamsQueue = Channel.CreateUnbounded<QuicStream>(new UnboundedChannelOptions()
+        {
+            SingleWriter = true
+        });
+
+        private int _disposed;
+        internal Http3WebtransportManager _WebtransportManager;
+
+        internal const string WebTransportProtocolValue = "webtransport";
+        internal const string VersionEnabledIndicator = "1";
+        internal const string SecPrefix = "sec-webtransport-http3-";
+        internal const string VersionHeaderPrefix = $"{SecPrefix}draft";
+        internal const string CurrentSuppportedVersion = $"{VersionHeaderPrefix}02";
+
+        internal Http3WebtransportSession(QuicStream connectStream, Http3WebtransportManager manager)
+        {
+            _WebtransportManager = manager;
+            _connectStream = connectStream;
+            _ = _connectStream.WritesClosed.ContinueWith(async t =>
```

Comment:
what will happen if a server closes only it's write side?

Reply:
As per RFC the session is closed when the " CONNECT stream is closed, either cleanly or abruptly, on **either** side;" so the session will be disposed
I will also add a complete reads